src/game/client/water/w3dwater.cpp

```diff
@@ -0,0 +1,2327 @@
+                            } else {
+                                m->velocity = m->velocity - g_theWriteableGlobalData->m_gravity * 3.0f;
+                            }
+
+                            m->height = m->height + m->velocity;
+
+                            if (GameMath::Fabs(m->height - m->preferred_height) >= 1.0f
+                                || GameMath::Fabs(m->velocity) >= 1.0f) {
+                                m_meshInMotion = true;
+                            } else {
+                                m->status &= ~1;
+                                m->height = m->preferred_height;
+                                m->velocity = 0.0f;
+                            }
+                        }
+
+                        m++;
+                    }
+                }
+            }
+        }
+        lastLogicFrame = frame;
+    }
+}
+
+void WaterRenderObjClass::Replace_Skybox_Texture(Utf8String const &oldname, Utf8String const &newname)
+{
+    ((GameAssetManager *)W3DAssetManager::Get_Instance())->Replace_Prototype_Texture(m_skyBox, oldname, newname);
+
+    if (m_skyBox && m_skyBox->Class_ID() == RenderObjClass::CLASSID_MESH) {
+        MeshClass *mesh = (MeshClass *)m_skyBox;
+        MaterialInfoClass *matinfo = mesh->Get_Material_Info();
+
+        for (int i = 0; i < matinfo->Texture_Count(); i++) {
+            if (matinfo->Peek_Texture(i)) {
+                matinfo->Peek_Texture(i)->Get_Texture_Filter()->Set_U_Address_Mode(
+                    TextureFilterClass::TEXTURE_ADDRESS_CLAMP);
+                matinfo->Peek_Texture(i)->Get_Texture_Filter()->Set_V_Address_Mode(
+                    TextureFilterClass::TEXTURE_ADDRESS_CLAMP);
+            }
+        }
+
+        Ref_Ptr_Release(matinfo);
+    }
+}
+
+void WaterRenderObjClass::Set_Time_Of_Day(TimeOfDayType tod)
+{
+#ifdef BUILD_WITH_D3D8
+    m_tod = tod;
+
+    if (m_waterType == WATER_TYPE_2_PVSHADER) {
+        Generate_Vertex_Buffer(15, 15, 24, true);
+    }
+#endif
+}
+
+void WaterRenderObjClass::Load_Setting(Setting *setting, TimeOfDayType tod)
+{
+    captainslog_assert(setting);
+
+    setting->sky_texture = W3DAssetManager::Get_Instance()->Get_Texture(g_waterSettings[tod].m_skyTextureFile);
+    setting->water_texture = W3DAssetManager::Get_Instance()->Get_Texture(g_waterSettings[tod].m_waterTextureFile);
+    setting->sky_texels_per_unit = g_waterSettings[tod].m_skyTexelsPerUnit;
+    SurfaceClass::SurfaceDescription surface_desc;
+    setting->water_texture->Get_Level_Description(surface_desc, 0);
+    setting->sky_texels_per_unit = setting->sky_texels_per_unit / surface_desc.width;
+    setting->water_repeat_count = g_waterSettings[tod].m_waterRepeatCount;
+    setting->u_scroll_per_ms = g_waterSettings[tod].m_uScrollPerMs;
+    setting->v_scroll_per_ms = g_waterSettings[tod].m_vScrollPerMs;
+
+    setting->vertex00_diffuse = g_waterSettings[tod].m_vertex00Diffuse.blue

```

Comment:
Add a TODO to change to use a packer function

Reply:
fixed.